### PR TITLE
MINIFICPP-926 Create nanofi tailfile processor for tailing file by configurable chunk size

### PR DESCRIPTION
**Pull request by msharee9**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by msharee9** [comment]
Created nanofi tailfile processor to tail file by configurable chunk size. Refactored existing code.
1. Created a file ecu.c that contains on_trigger_<processor> callbacks.
2. Separated out the callback implementations from the file containing main function for each processor.
The callbacks can be invoked in unit tests to test out the processor as a whole.
3. Renamed the previous mixed mode tail file implementation to log_aggregator.

File: nanofi/ecu/log_aggregator.c [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    char * port_str = argv[5];
+
+    if (access(file, F_OK) == -1) {
+        printf("Error: %s doesn't exist!\n", file);
+        exit(1);
+    }
+
+    int i;
+    for (i = 0; i < strlen(delimiter); ++i) {
+        printf("%d - %c\n", i, delimiter[i]);
```
[the remark this answers is not in the document]
Comment (msharee9) [reply nanofi/ecu/log_aggregator.c]: yes.

File: nanofi/src/api/ecu.c [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    if (get_property(ctx, "file_path", file_path, sizeof(file_path)) != 0) {
+        return;
+    }
+
+    if (get_property(ctx, "chunk_size", chunk_size, sizeof(chunk_size)) != 0) {
+        return;
+    }
+
+    errno = 0;
+    unsigned long chunk_size_value = strtol(chunk_size, NULL, 10);
```
[the remark this answers is not in the document]
Comment (msharee9) [reply nanofi/src/api/ecu.c]: good point.

File: nanofi/src/api/ecu.c [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+#include <string.h>
+#include <errno.h>
+#include <limits.h>
+#include <signal.h>
+#include <sys/stat.h>
+
+nifi_instance * instance = NULL;
+standalone_processor * proc = NULL;
+flow_file_list ff_list;
+int curr_offset = 0;
```
[the remark this answers is not in the document]
Comment (msharee9) [reply nanofi/src/api/ecu.c]: good catch.

**Comment by msharee9** [comment]
> If I understand correctly this PR contains two modes (chunked and mixed) and does not yet contain the delimited mode. Do you plan on doing that in a separate PR?
> I think if we want to include a mixed mode, both the delimiter and the maximum chunk size should be specifiable (now only the delimiter can be specified, the chunk size is fixed).
> 
> Please also see my inline comments.

@bakaid Yes, I plan to implement the delimited only mode in a separate PR that is MINIFICPP-927.
I agree with your point to make chunk size configurable as well for the mixed mode.

File: nanofi/src/core/string_utils.c [hunk 4]
```diff
@@ -20,67 +20,145 @@
...
+
+        if (!to->tail) return;
+
+        to->tail->next = from->head;
+        to->tail = from->tail;
+        to->size += from->size;
+    }
+}
+
+token_list tokenize_string(const char * begin, char delim) {
```
Comment (arpadboda) [inline nanofi/src/core/string_utils.c]: Wonder if we could make this return utlist as well

File: nanofi/tests/CLogAggregatorTests.cpp [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+std::string join_strings(const std::vector<std::string>& strings, const std::string& token) {
+    if (strings.empty()) {
+        return "";
+    }
+
+    if (strings.size() == 1) {
+        return strings.at(0);
+    }
+    const auto& initialValue = strings.at(0);
+    return std::accumulate(strings.begin()+1, strings.end(), initialValue,
```
Comment (arpadboda) [inline nanofi/tests/CLogAggregatorTests.cpp]: You can use StringUtils::join now

File: nanofi/src/core/string_utils.c (hunk 4 above)
Comment (msharee9) [reply nanofi/src/core/string_utils.c]: This function returns a list of tokens (strings) when called. However, it is also used as part of tokenizing for log aggregate usage. The log aggregate usage has special requirements, therefore this function is used in conjunction with other function in the same file. In order to tailor utlist to fulfill those special requirements, some extra functions will be needed anyway.

File: nanofi/include/core/file_utils.h [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+ * @param file the path to the file to tail
+ * @param delim the delimiter character
+ * @param curr_offset the offset in the file to tail from.
+ * For eg. To tail from beginning of the file curr_offset = 0
+ * @param instance the nifi_instance
+ * @param proc the standalone_processor
+ * @return a list of flow file info containing list of flow file records
+ * and the current offset in the file
+ */
+flow_file_info log_aggregate(const char * file_path, char delim, uint64_t curr_offset, nifi_instance * instance, standalone_processor * proc);
```
[the remark this answers is not in the document]
Comment (msharee9) [reply nanofi/include/core/file_utils.h]: Segregated this out of this file.

File: nanofi/src/core/file_utils.c [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    if (stat(path, &dir_stat) < 0) {
+        return 0;
+    }
+    return S_ISDIR(dir_stat.st_mode);
+}
+
+char * concat_path(const char * parent, const char * child) {
+    char * path = (char *)malloc((strlen(parent) + strlen(child) + 2) * sizeof(char));
+    strcpy(path, parent);
+    strcat(path, "/");
```
[the remark this answers is not in the document]
Comment (msharee9) [reply nanofi/src/core/file_utils.c]: Got it.

File: nanofi/src/core/flowfiles.c [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+#include <stdlib.h>
+
+void add_flow_file_record(flow_file_list * ff_list, flow_file_record * record) {
+    if (!ff_list || !record) return;
+
+    struct flow_file_list_node * new_node = (struct flow_file_list_node *)malloc(sizeof(struct flow_file_list_node));
+    new_node->ff_record = record;
+    new_node->next = NULL;
+
+    if (!ff_list->head || !ff_list->tail) {
```
[the remark this answers is not in the document]
Comment (msharee9) [reply nanofi/src/core/flowfiles.c]: Yes this can be done.

File: nanofi/src/api/nanofi.cpp [hunk 9]
```diff
@@ -263,10 +263,9 @@ flow_file_record * generate_flow_file(nifi_instance * instance, standalone_proce
     ffr->keepContent = 1;
     auto ff_content_repo_ptr = (static_cast<std::shared_ptr<minifi::core::ContentRepository>*>(ffr->crp));
     auto claim = std::make_shared<minifi::ResourceClaim>(*ff_content_repo_ptr);
-    const char * full_path = claim->getContentFullPath().c_str();
```
[the remark this answers is not in the document]
Comment (msharee9) [reply nanofi/src/api/nanofi.cpp]: I am using create_processor function to create a processor. This function creates a flow, a plan and links them together. I need the UUID of the processor in order to get the plan and in turn the flow file ptr.

File: nanofi/include/api/ecu.h [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+#define NANOFI_INCLUDE_API_ECU_H_
+
+#include <signal.h>
+#include "api/nanofi.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+extern nifi_instance * instance;
```
[the remark this answers is not in the document]
Comment (msharee9) [reply nanofi/include/api/ecu.h]: I will refactor this as part of MINIFICPP-927 PR.

**Comment by arpadboda** [comment]
@msharee9 : I think this one can be closed now. If you agree, please close it!

**Pull request closed without merging**